### PR TITLE
Add Travis CI check for line lengths

### LINKED ISSUE #42: Line wrap convention?
What's the line wrap convention? I initially guessed 80 chars max, but then it seems like a) a few lines overflow (I'd guess these were oversights), b) many lines wrap before they need to (I'd guess intentionally). I can't figure out the logic for b) in particular, since no consistent scheme seems to be used.

Personally, I'm not a big fan of hard wrapping in markup files (or code, to a lesser extent, though I know that's more contentious!). Should we be doing wrapping at all? If yes, let's document a consistent style, I say.

### PR DESCRIPTION
**Pull request by alexreg**
Fixes #42.

I don't know if you want this as a Git commit hook too, but maybe you can add that if you like.

### CODE REVIEW

File: ci/check_line_lengths.sh [hunk 1]
```diff
@@ -0,0 +1,20 @@
+#!/bin/bash
+
+echo "Checking line lengths in all source files <= $MAX_LINE_LENGTH chars..."
+
+echo "Offending files and lines:"
+(( success = 1 ))
+for file in src/*.md ; do
```
Comment (mark-i-m) [inline ci/check_line_lengths.sh]: I think we also need to check subdirectories (e.g. `test`)

Comment (alexreg) [reply ci/check_line_lengths.sh]: Yeah. We have a problem though: in tables, we can't just wrap at 80 chars. Hrmm.

**Comment by mark-i-m** [comment]
Hmm... maybe add a list of exceptions? I think `glossary.md` and `code_index.md` are the only offending files, right?

**Comment by mark-i-m** [comment]
And maybe lines that have URLs in them? (e.g. maybe just grepping for `https://` in the line and allowing if that is found)

**Comment by alexreg** [comment]
Okay, done! I made the line length checker a bit smarter rather than just ignoring those files. It does ignore certain lines though now.

**Comment by alexreg** [comment]
Ah rubbish, you merged... rebase time.

**Comment by alexreg** [comment]
So, fixing the offending lines really is a pain in the arse, due to the knock-on effect on subsequent lines... this is one of the reasons I hate hard wrapping. :P Some of the lines are fixed now, but not sure when I'll get around to fixing the others. If you have a patch, feel free to submit, and we can get this PR merged.

**Comment by mark-i-m** [comment]
@alexreg Maybe the correct approach would be to comment out the check by Travis for the moment and add some sort of script wrapping fmt. Then just before we merge a PR we ask the OP to room the script on it?

We can reenable the check when all files are passing?

File: src/SUMMARY.md [hunk 2]
```diff
@@ -6,7 +6,8 @@
 - [The compiler testing framework](./tests/intro.md)
     - [Running tests](./tests/running.md)
     - [Adding new tests](./tests/adding.md)
-    - [Using `compiletest` + commands to control test execution](./compiletest.md)
+    - [Using `compiletest` + commands to control test
```
Comment (mark-i-m) [inline src/SUMMARY.md]: Perhaps we can just shorten this to `compiletest`

**Comment by mark-i-m** [comment]
Ah, I see :/

It seems strange that there isn't a good standard solution to this.

How about this:
- we add a whitelist of files that are temporarily allowed to have long lines, and add all existing files (that you haven't already fixed)
- we merge this PR so that new files get the check
- we create a tracking issue and incrementally pick off files from the list until we can remove the whitelist

What do you think?

**Comment by nikomatsakis** [comment]
I will happily fix the long lines, but when I try to build this PR locally, I get this:

```bash 
> mdbook build
2018-03-02 08:42:08 [INFO] (mdbook::book): Book building has started
2018-03-02 08:42:08 [INFO] (mdbook::book): Running the html backend
2018-03-02 08:42:08 [INFO] (mdbook::book): Running the linkcheck backend
2018-03-02 08:42:08 [INFO] (mdbook::renderer): Invoking the "linkcheck" renderer
There was 1 broken link:
how-to-build-and-run.md#140: "./running-tests.html" doesn't exist
2018-03-02 08:42:08 [ERROR] (mdbook::renderer): Renderer exited with non-zero return code.
2018-03-02 08:42:08 [ERROR] (mdbook::utils): Error: Rendering failed
2018-03-02 08:42:08 [ERROR] (mdbook::utils):    Caused By: The "linkcheck" renderer failed
```

I guess it needs a rebase? @alexreg -- do you object to me doing that?

**Comment by nikomatsakis** [comment]
That said, personally I don't know that we need to enforce a lack of long lines. But I'm always inclined towards a "laissez faire" attitude when it comes to style issues I guess.

**Comment by alexreg** [comment]
@nikomatsakis Yeah, probably just a rebase needed. Doesn't seem like anything to do with my edits. Personally, I'm in favour of not hard wrapping any lines (and thus saving future pain), but if we're going to wrap, we might as well enforce a consistent style. So please feel free to go ahead with that!

**Comment by nikomatsakis** [comment]
Rebased -- but @alexreg will `mdbook build` fail locally if there are long lines? If not, how do I run this locally?

**Comment by nikomatsakis** [comment]
Running `find . -name '*md' | xargs bash ci/check_line_length.sh` generates all kinds of crazy errors on a mac. Probably this assumes linux bash?

**Comment by nikomatsakis** [comment]
Never mind, running exactly the command from travis seems to work.

**Comment by alexreg** [comment]
@nikomatsakis Glad you solved it already. I use Bash 4 on Mac (not the OS’s Bash 3), so it works on that.

File: .travis.yml [hunk 3]
```diff
@@ -1,6 +1,8 @@
 language: rust
 cache:
 - cargo
+before_install:
+- MAX_LINE_LENGTH=80 bash ci/check_line_lengths.sh src/**/*.md
```
Comment (mark-i-m) [inline .travis.yml]: This seems to only be running on the src/test/ directory. Perhaps this should be changed to

```
- MAX_LINE_LENGTH=80 bash ci/check_line_lengths.sh src/*.md src/**/*.md 
```

Comment (alexreg) [reply .travis.yml]: Yeah, actually, `src/**.md` may be what we want...?

Comment (mark-i-m) [reply .travis.yml]: Nope, it looks like that doesn't get the tests/ directory.

**Comment by alexreg** [comment]
There might be an `shopt` we need to enable?

**Comment by alexreg** [comment]
Just tried it on my computer, and `src/**.md` definitely works!

**Comment by mark-i-m** [comment]
But does it get the `tests/` directory?

The easiest thing might be `MAX_LINE_LENGTH=80 ./ci/check_line_lengths.sh $(find -type f -name '*.md')`

**Comment by alexreg** [comment]
Sure it does. This is standard globbing on Bash. Always works. Maybe Travis has a really old version of Bash...

**Comment by mark-i-m** [comment]
?? It doesn't work on my local machine either? Am I missing something?

**Comment by alexreg** [comment]
Version of Bash? Output of `shopt`?

**Comment by mark-i-m** [comment]
```
$ shopt
autocd         	off
cdable_vars    	off
cdspell        	off
checkhash      	off
checkjobs      	off
checkwinsize   	on
cmdhist        	on
compat31       	off
compat32       	off
compat40       	off
compat41       	off
compat42       	off
complete_fullquote	on
direxpand      	off
dirspell       	off
dotglob        	off
execfail       	off
expand_aliases 	on
extdebug       	off
extglob        	on
extquote       	on
failglob       	off
force_fignore  	on
globstar       	off
globasciiranges	off
gnu_errfmt     	off
histappend     	on
histreedit     	off
histverify     	off
hostcomplete   	off
huponexit      	off
interactive_comments	on
lastpipe       	off
lithist        	off
login_shell    	off
mailwarn       	off
no_empty_cmd_completion	off
nocaseglob     	off
nocasematch    	off
nullglob       	off
progcomp       	on
promptvars     	on
restricted_shell	off
shift_verbose  	off
sourcepath     	on
xpg_echo       	off

$ bash --version
GNU bash, version 4.3.48(1)-release (x86_64-pc-linux-gnu)
Copyright (C) 2013 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>

This is free software; you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
```

**Comment by alexreg** [comment]
I just double-checked, and both `src/**.md` and the original `src/**/*.md` do the job perfectly.

It turns out that the `globstar` shell option wasn't on in your shell. (It's good common practice to have it on, but it's not default nonetheless, probably for compatibility reasons. I'm setting it in my bashrc it seems.) Anyway, I've now modified the script to turn it on manually. Let me know if that does the tricky for you!

**Comment by mark-i-m** [comment]
Thanks! I will be a bit busy until Monday, but I plan to come back and help fix long lines next week...

Do you know why Travis is passing?

**Comment by alexreg** [comment]
@mark-i-m Oddly, it seems to only be checking the files in `tests` still. Don't ask why.

**Comment by mark-i-m** [comment]
@alexreg Oh... ok... that's a bit strange. Perhaps for now we can just use `src/*.md src/**/*.md`?

**Comment by alexreg** [comment]
> @alexreg Oh... ok... that's a bit strange. Perhaps for now we can just use src/*.md src/**/*.md?

Very strange. Probably a non-standard shopt on the CI server. I recommend someone check into that, who knows how Travis CI works.

**Comment by ehuss** [comment]
> Very strange. Probably a non-standard shopt on the CI server. I recommend someone check into that, who knows how Travis CI works.

The call to `shopt` is inside the shell script, but the expansion has already occurred in the `travis.yml` script before `check_line_lengths.sh` is called.  You need to call `shopt` from within `travis.yml` before calling the script.

**Comment by mark-i-m** [comment]
Lol, good call :+1:

**Comment by alexreg** [comment]
@ehuss D'oh. I made that "fix" a bit too quickly.

**Comment by alexreg** [comment]
Failures. \o/

**Comment by mark-i-m** [comment]
@alexreg If you want, I can help clean up all of the offending lines. What would be the best way to do that?

**Comment by alexreg** [comment]
@mark-i-m, Thanks. I think it would be easiest just to run the `check_line_lengths.sh` script locally and go fix the lines by number. @nikomatsakis also offered to help with this I know. Now that it's in master, do you want to just tackle it there? :-)

**Comment by mark-i-m** [comment]
Oh, lol. I should have waited for your answer :) I just made a PR on your repo. I would like to not merge a travis-failing PR to master if possible.

**Comment by alexreg** [comment]
Fair enough hah. Are we all ready to merge now?

**Comment by mark-i-m** [comment]
Ok, I think I got all the rest :)

**Comment by alexreg** [comment]
Great!

**Comment by mark-i-m** [comment]
:interrobang: 

```
$ MAX_LINE_LENGTH=80 ./ci/check_line_lengths.sh src/*.md
Checking line lengths in all source files <= 80 chars...
Offending files and lines:
src/about-this-guide.md
src/appendix-background.md
src/appendix-code-index.md
src/appendix-glossary.md
src/appendix-stupid-stats.md
src/compiletest.md
src/const-eval.md
src/conventions.md
src/high-level-overview.md
src/hir.md
src/how-to-build-and-run.md
src/incrcomp-debugging.md
src/incremental-compilation.md
src/macro-expansion.md
src/method-lookup.md
src/mir-borrowck.md
src/mir-construction.md
src/miri.md
src/mir.md
src/mir-optimizations.md
src/mir-passes.md
src/mir-regionck.md
src/mir-visitor.md
src/name-resolution.md
src/param_env.md
src/query.md
src/rustc-driver.md
src/rustdoc.md
src/SUMMARY.md
src/the-parser.md
src/trait-caching.md
src/trait-hrtb.md
src/trait-resolution.md
src/traits-associated-types.md
src/traits-bibliography.md
src/traits-canonicalization.md
src/traits-canonical-queries.md
src/traits-goals-and-clauses.md
src/traits-implied-bounds.md
src/traits-lowering-module.md
src/traits-lowering-rules.md
src/traits-lowering-to-logic.md
src/traits.md
src/trait-specialization.md
src/traits-regions.md
src/traits-slg.md
src/traits-wf.md
src/trans.md
src/ty.md
src/type-checking.md
src/type-inference.md
src/variance.md
src/walkthrough.md
0 offending lines found.
```

**Comment by mark-i-m** [comment]
@alexreg Could you rebase off master please? Maybe there are some additions we missed.

**Comment by mark-i-m** [comment]
Thanks @alexreg!

**Comment by alexreg** [comment]
@mark-i-m Sure. Likewise thanks for helping out and getting this merged! And to @nikomatsakis too.
